File: Algoritmos y Programacion II/AYP2go/Guias/TDAsPilasColasListas/ej6.go
// Dada una pila de enteros, escribir una función que determine si sus elementos están ordenados de manera ascendente. 
// Una pila de enteros está ordenada de manera ascendente si, en el sentido que va desde el tope de la pila hacia el 
// resto de elementos, cada elemento es menor al elemento que le sigue. La pila debe quedar en el mismo estado que al 
// invocarse la función. 
// Indicar y justificar el orden del algoritmo propuesto.

package main

import (
	"fmt"
	tda "tdas/pila"
)

func main() {

}

func OrdenadosAscendentes(pila tda.Pila[int]) bool {
	pilaAux := tda.CrearPilaDinamica[int]()
	anterior := pila.Desapilar() // tope
	pilaAux.Apilar(anterior)
	esAsc := false

	for ! pila.EstaVacia() {
		elem := pila.Desapilar()
		pilaAux.Apilar(elem)
		if elem <= anterior {
			esAsc = true
		}
		anterior = elem
	}

	for ! pilaAux.EstaVacia() {
		// para reestablecer la pila
		pila.Apilar(pilaAux.Desapilar())
	}

	if esAsc {
		return false
	}
	return true
} 

File: Algoritmos y Programacion I/Ejercicios - Guia/14. Pilas y Colas/EJ6.py
from modelo_Pila import Pila

def validar(expresion):
	pila = Pila()
	for c in expresion:
		if c in '([{':
			pila.apilar(c)
		elif c in ')]}':
			if pila.esta_vacia():
				return False
			tope = pila.desapilar()
			if c == ')' and tope != '(':
				return False
			if c == ']' and tope != '[':
				return False
			if c == '}' and tope != '{':
				return False
	return pila.esta_vacia()
